<a href="https://colab.research.google.com/github/karim-mttk/machine-learning-basics/blob/main/ML_Project2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Project 2: Decision Trees
## Objectives:
* Building Decision Trees with `sklearn` 

3 types of dataset:
* Training dataset
* Validation dataset
* Test set

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

from sklearn.dummy import DummyClassifier
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.linear_model import LinearRegression, LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline, make_pipeline
from sklearn.tree import DecisionTreeClassifier, DecisionTreeRegressor, export_graphviz

In [3]:
url = "https://raw.githubusercontent.com/karim-mttk/machine-learning-basics/main/data/grade-toy-classification.csv"
classi_df = pd.read_csv(url)

* Setting the target and features: 

In [7]:
X = classi_df.drop(columns =["quiz2"])
y = classi_df["quiz2"]
X_bin = X.copy()
columns = ["lab1", "lab2", "lab3", "lab4", "quiz1"]
for col in columns:
  X_bin[col] = X_bin[col].apply(lambda x: 1 if x >= 90 else 0)
X_bin.head()

ml_experience  class_attendance  lab1  lab2  lab3  lab4  quiz1
0              1                 1     1     1     0     1      1
1              1                 0     1     1     0     0      1
2              0                 0     0     0     0     0      0
3              0                 1     1     1     1     1      0
4              0                 1     0     0     1     1      0

In [8]:
y.head()

0        A+
1    not A+
2    not A+
3        A+
4        A+
Name: quiz2, dtype: object

* Using `DummyClassifier` on quiz2 grade prediction:

In [11]:
#1. create dummy classifier model
dummy_clf = DummyClassifier(strategy="most_frequent")
#2. Fit the model
dummy_clf.fit(X_bin, y)
#3. Score the model
dummy_clf.score(X_bin, y)

0.5238095238095238

* Using `DecisionTreeClassifier` on quiz2:

In [13]:
from sklearn.tree import DecisionTreeClassifier

#1. Create a decision tree
model = DecisionTreeClassifier()
#2. fit the decision tree
model.fit(X_bin, y)
#3. score the model
model.score(X_bin, y)


0.9047619047619048

From the two models we can see that the `DecisionTreeClassifier` has a higher accuracy of 0.90 compared to `DummyClassfier` of 0.52